lms/services/digest.py

```diff
@@ -0,0 +1,267 @@
+from dataclasses import dataclass, field
+from functools import lru_cache
+from typing import List, Optional
+
+from sqlalchemy import func, select
+
+from lms.models import (
+    Assignment,
+    AssignmentGrouping,
+    AssignmentMembership,
+    Course,
+    Grouping,
+    LTIRole,
+    User,
+)
+from lms.services.h_api import HAPI
+from lms.services.mailchimp import EmailRecipient, EmailSender, MailchimpService
+
+
+class DigestService:
+    def __init__(
+        self, db, helper, h_api, mailchimp_service, sender
+    ):  # pylint:disable=too-many-arguments
+        self.db = db
+        self.helper = helper
+        self.h_api = h_api
+        self.mailchimp_service = mailchimp_service
+        self.sender = sender
+
+    def send_instructor_email_digests(
+        self, h_userids, since, until, override_to_email=None
+    ):
+        annotations = self.h_api.get_annotations(h_userids, since, until)
+        unified_courses = self.helper.unified_courses(annotations)
+
+        for h_userid in h_userids:
+            unified_user = UnifiedUser.make(self.db, h_userid)
+            digest = self.helper.instructor_digest(unified_user, unified_courses)
+
+            if not digest["num_annotations"]:
+                continue
+
+            if override_to_email is None:
+                to_email = unified_user.email
+            else:
+                to_email = override_to_email
+
+            self.mailchimp_service.send_template(
+                "instructor-email-digest",
+                self.sender,
+                recipient=EmailRecipient(to_email, unified_user.display_name),
+                template_vars=digest,
+            )
+
+
+@dataclass
+class UnifiedUser:
+    """All User's for a given h_userid, unified across all ApplicationInstance's."""
```

Comment:
I think Jon wanted me to get rid of this `UnifiedUser` class and the `UnifiedCourse` one below but I found it really hard to write the code without these model classes. The digests algorithm really wants to speak in terms of unified users and groups that aren't split by instance, and I found it really hard to write the algorithm without these nouns (or by using dicts and lists etc).

It definitely feels like these two model classes don't belong here as private helpers of this service and they're a big reason for the high line count in this PR.

But I think the service code here is struggling with our lack of unified user and course models. We have some missing models and/or service methods. So perhaps we can look into factoring these helpers away into models or service methods in future. But right now I don't want to take the time to figure that out.